Use dataclass properties in yeelight discovery

epenet opened the pull request:
## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->
Use dataclass properties in `yeelight` discovery.

Linked to #60540

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [ ] New feature (which adds functionality to an existing integration)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [x] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #
- This PR is related to issue: 
- Link to documentation pull request: 

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [ ] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] The code has been formatted using Black (`black --fast homeassistant tests`)
- [ ] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [ ] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.
- [ ] Untested files have been added to `.coveragerc`.

The integration reached or maintains the following [Integration Quality Scale][quality-scale]:
<!--
  The Integration Quality Scale scores an integration on the code quality
  and user experience. Each level of the quality scale consists of a list
  of requirements. We highly recommend getting your integration scored!
-->

- [ ] No score or internal
- [ ] 🥈 Silver
- [ ] 🥇 Gold
- [ ] 🏆 Platinum

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [x] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io

epenet committed:
Use dataclass properties in yeelight discovery

epenet commented:
I did not update `async_step_ssdp` yet because I am not 100% sure if these two fields come from the SSDP headers or from the UPnP information:
https://github.com/home-assistant/core/blob/0a1f73e3f7feeff6590cc3038db612195d3dcc27/homeassistant/components/yeelight/config_flow.py#L81-L84

For the tests, I assumed it was UPnP but if I made a mistake then I can correct it in this PR :
https://github.com/home-assistant/core/blob/0a1f73e3f7feeff6590cc3038db612195d3dcc27/tests/components/yeelight/test_config_flow.py#L55-L59

If you do not have the opportunity to check it, then I suggest that we merge it as it is to avoid blocking #60561, and let the warning kick in when #60540 is merged.

bdraco commented:
> I did not update `async_step_ssdp` yet because I am not 100% sure if these two fields come from the SSDP headers or from the UPnP information:

```
2021-11-30 06:16:05 WARNING (MainThread) [homeassistant.components.yeelight.config_flow] dinfo: {'Cache-Control': 'max-age=3600', 'Date': '', 'Ext': '', 'Server': 'POSIX UPnP/1.0 YGLC/1', 'id': '0x00000000158d5a91', 'model': 'ceiling10', 'fw_ver': '49', 'support': 'get_prop set_default set_power toggle set_bright set_scene cron_add cron_get cron_del start_cf stop_cf set_ct_abx set_name set_adjust adjust_bright adjust_ct bg_set_rgb bg_set_hsv bg_set_ct_abx bg_start_cf bg_stop_cf set_scene_bundle bg_set_default bg_set_power bg_set_bright bg_set_adjust bg_adjust_bright bg_adjust_color bg_adjust_ct bg_toggle dev_toggle', 'power': 'on', 'bright': '100', 'color_mode': '2', 'ct': '2801', 'rgb': '0', 'hue': '0', 'sat': '0', 'name': '', '_location_original': 'yeelight://192.168.107.53:55443', 'location': 'yeelight://192.168.107.53:55443', '_timestamp': datetime.datetime(2021, 11, 30, 6, 16, 4, 56326), '_host': '192.168.107.53', '_port': 49154, '_source': <SsdpSource.SEARCH: 'search'>}
2021-11-30 06:16:05 WARNING (MainThread) [homeassistant.components.yeelight.config_flow] dinfo: {'Cache-Control': 'max-age=3600', 'Date': '', 'Ext': '', 'Server': 'POSIX UPnP/1.0 YGLC/1', 'id': '0x000000000ed53acf', 'model': 'colorb', 'fw_ver': '10', 'support': 'get_prop set_default set_power toggle set_bright set_scene cron_add cron_get cron_del start_cf stop_cf set_ct_abx adjust_ct set_name set_adjust adjust_bright adjust_color set_rgb set_hsv set_music udp_sess_new udp_sess_keep_alive udp_chroma_sess_new', 'power': 'on', 'bright': '100', 'color_mode': '2', 'ct': '2801', 'rgb': '8388863', 'hue': '270', 'sat': '100', 'name': '', '_location_original': 'yeelight://192.168.106.16:55443', 'location': 'yeelight://192.168.106.16:55443', '_timestamp': datetime.datetime(2021, 11, 30, 6, 16, 4, 63776), '_host': '192.168.106.16', '_port': 49153, '_source': <SsdpSource.SEARCH: 'search'>}
```

epenet committed:
Use ssdp_headers for async_step_ssdp

epenet commented:
Thanks @bdraco, I have moved this to ssdp_headers.
However the tests now fail with `AttributeError: 'dict' object has no attribute 'ssdp_headers'`

I assume that I have missed a fixture somewhere when I changed all the tests but I can't find it :(

epenet committed:
Fix tests

epenet commented:
It was an issue in `_patch_discovery`. That is fixed now.
Hassfest is still complaining about the ssdp dependency in hassfest:
```
Integrations: 1015
Invalid integrations: 1

Integration yeelight:
Error: R] [DEPENDENCIES] Using component ssdp but it's not in 'dependencies' or 'after_dependencies'
```

bdraco commented on tests/components/yeelight/__init__.py:
```diff
@@ -180,7 +180,14 @@ def _patch_discovery(no_device=False, capabilities=None):
     def _generate_fake_ssdp_listener(*args, **kwargs):
         return _patched_ssdp_listener(
-            None if no_device else capabilities or CAPABILITIES,
+            None
+            if no_device
+            else ssdp.SsdpServiceInfo(
+                ssdp_usn="mock_usn",
+                ssdp_st="mock_st",
+                upnp={},
+                ssdp_headers=capabilities or CAPABILITIES,
+            ),
```
This should be unpacked into an `if` block since it spans multiple lines

epenet replied:
Done

epenet committed:
Adjust tests

epenet committed:
Add after_dependencies

epenet commented:
I have added `"after_dependencies": ["ssdp"]` to the manifest and now this passes.

bdraco commented:
@epenet 
Saw this trace this morning on my integration branch
```
2021-12-01 05:59:03 ERROR (MainThread) [homeassistant] Error doing job: Task exception was never retrieved:   File "/Users/bdraco/home-assistant/venv/bin/hass", line 8, in <module>
    sys.exit(main())
  File "/Users/bdraco/home-assistant/venv/lib/python3.9/site-packages/homeassistant/__main__.py", line 318, in main
    exit_code = runner.run(runtime_conf)
  File "/Users/bdraco/home-assistant/venv/lib/python3.9/site-packages/homeassistant/runner.py", line 121, in run
    return loop.run_until_complete(setup_and_run_hass(runtime_config))
  File "/opt/homebrew/Cellar/python@3.9/3.9.6/Frameworks/Python.framework/Versions/3.9/lib/python3.9/asyncio/base_events.py", line 629, in run_until_complete
    self.run_forever()
  File "/opt/homebrew/Cellar/python@3.9/3.9.6/Frameworks/Python.framework/Versions/3.9/lib/python3.9/asyncio/base_events.py", line 596, in run_forever
    self._run_once()
  File "/opt/homebrew/Cellar/python@3.9/3.9.6/Frameworks/Python.framework/Versions/3.9/lib/python3.9/asyncio/base_events.py", line 1882, in _run_once
    handle._run()
  File "/opt/homebrew/Cellar/python@3.9/3.9.6/Frameworks/Python.framework/Versions/3.9/lib/python3.9/asyncio/events.py", line 80, in _run
    self._context.run(self._callback, *self._args)
  File "/Users/bdraco/home-assistant/venv/lib/python3.9/site-packages/homeassistant/helpers/event.py", line 1273, in run_action
    hass.async_run_hass_job(job, utc_point_in_time)
  File "/Users/bdraco/home-assistant/venv/lib/python3.9/site-packages/homeassistant/core.py", line 433, in async_run_hass_job
    hassjob.target(*args)
  File "/Users/bdraco/home-assistant/venv/lib/python3.9/site-packages/homeassistant/components/yeelight/scanner.py", line 160, in _async_start_flow
    asyncio.create_task(
  File "/opt/homebrew/Cellar/python@3.9/3.9.6/Frameworks/Python.framework/Versions/3.9/lib/python3.9/asyncio/tasks.py", line 361, in create_task
    task = loop.create_task(coro)
Traceback (most recent call last):
  File "/Users/bdraco/home-assistant/venv/lib/python3.9/site-packages/homeassistant/data_entry_flow.py", line 203, in async_init
    flow, result = await task
  File "/Users/bdraco/home-assistant/venv/lib/python3.9/site-packages/homeassistant/data_entry_flow.py", line 230, in _async_init
    result = await self._async_handle_step(flow, flow.init_step, data, init_done)
  File "/Users/bdraco/home-assistant/venv/lib/python3.9/site-packages/homeassistant/data_entry_flow.py", line 325, in _async_handle_step
    result: FlowResult = await getattr(flow, method)(user_input)
  File "/Users/bdraco/home-assistant/venv/lib/python3.9/site-packages/homeassistant/components/yeelight/config_flow.py", line 83, in async_step_ssdp
    self._discovered_ip = urlparse(discovery_info.ssdp_headers["location"]).hostname
AttributeError: 'CaseInsensitiveDict' object has no attribute 'ssdp_headers'
```

bdraco commented:
```
2021-12-01 06:16:41 DEBUG (MainThread) [homeassistant.components.yeelight.scanner] Discovered via SSDP: {'Cache-Control': 'max-age=3600', 'Date': '', 'Ext': '', 'Server': 'POSIX UPnP/1.0 YGLC/1', 'id': '0x00000000158d5a91', 'model': 'ceiling10', 'fw_ver': '49', 'support': 'get_prop set_default set_power toggle set_bright set_scene cron_add cron_get cron_del start_cf stop_cf set_ct_abx set_name set_adjust adjust_bright adjust_ct bg_set_rgb bg_set_hsv bg_set_ct_abx bg_start_cf bg_stop_cf set_scene_bundle bg_set_default bg_set_power bg_set_bright bg_set_adjust bg_adjust_bright bg_adjust_color bg_adjust_ct bg_toggle dev_toggle', 'power': 'off', 'bright': '100', 'color_mode': '2', 'ct': '2801', 'rgb': '0', 'hue': '0', 'sat': '0', 'name': '', '_location_original': 'yeelight://192.168.107.53:55443', 'location': 'yeelight://192.168.107.53:55443', '_timestamp': datetime.datetime(2021, 12, 1, 6, 16, 41, 467365), '_host': '192.168.107.53', '_port': 49154, '_source': <SsdpSource.SEARCH: 'search'>}
2021-12-01 06:16:41 DEBUG (MainThread) [homeassistant.components.yeelight.scanner] Yeelight discovered with {'Cache-Control': 'max-age=3600', 'Date': '', 'Ext': '', 'Server': 'POSIX UPnP/1.0 YGLC/1', 'id': '0x00000000158d5a91', 'model': 'ceiling10', 'fw_ver': '49', 'support': 'get_prop set_default set_power toggle set_bright set_scene cron_add cron_get cron_del start_cf stop_cf set_ct_abx set_name set_adjust adjust_bright adjust_ct bg_set_rgb bg_set_hsv bg_set_ct_abx bg_start_cf bg_stop_cf set_scene_bundle bg_set_default bg_set_power bg_set_bright bg_set_adjust bg_adjust_bright bg_adjust_color bg_adjust_ct bg_toggle dev_toggle', 'power': 'off', 'bright': '100', 'color_mode': '2', 'ct': '2801', 'rgb': '0', 'hue': '0', 'sat': '0', 'name': '', '_location_original': 'yeelight://192.168.107.53:55443', 'location': 'yeelight://192.168.107.53:55443', '_timestamp': datetime.datetime(2021, 12, 1, 6, 16, 41, 467365), '_host': '192.168.107.53', '_port': 49154, '_source': <SsdpSource.SEARCH: 'search'>}

```

bdraco commented:
```
2021-12-01 06:16:41 DEBUG (MainThread) [homeassistant.components.yeelight.scanner] Discovered via SSDP: {'Cache-Control': 'max-age=3600', 'Date': '', 'Ext': '', 'Server': 'POSIX UPnP/1.0 YGLC/1', 'id': '0x00000000158d5a91', 'model': 'ceiling10', 'fw_ver': '49', 'support': 'get_prop set_default set_power toggle set_bright set_scene cron_add cron_get cron_del start_cf stop_cf set_ct_abx set_name set_adjust adjust_bright adjust_ct bg_set_rgb bg_set_hsv bg_set_ct_abx bg_start_cf bg_stop_cf set_scene_bundle bg_set_default bg_set_power bg_set_bright bg_set_adjust bg_adjust_bright bg_adjust_color bg_adjust_ct bg_toggle dev_toggle', 'power': 'off', 'bright': '100', 'color_mode': '2', 'ct': '2801', 'rgb': '0', 'hue': '0', 'sat': '0', 'name': '', '_location_original': 'yeelight://192.168.107.53:55443', 'location': 'yeelight://192.168.107.53:55443', '_timestamp': datetime.datetime(2021, 12, 1, 6, 16, 41, 467365), '_host': '192.168.107.53', '_port': 49154, '_source': <SsdpSource.SEARCH: 'search'>}
2021-12-01 06:16:41 DEBUG (MainThread) [homeassistant.components.yeelight.scanner] Yeelight discovered with {'Cache-Control': 'max-age=3600', 'Date': '', 'Ext': '', 'Server': 'POSIX UPnP/1.0 YGLC/1', 'id': '0x00000000158d5a91', 'model': 'ceiling10', 'fw_ver': '49', 'support': 'get_prop set_default set_power toggle set_bright set_scene cron_add cron_get cron_del start_cf stop_cf set_ct_abx set_name set_adjust adjust_bright adjust_ct bg_set_rgb bg_set_hsv bg_set_ct_abx bg_start_cf bg_stop_cf set_scene_bundle bg_set_default bg_set_power bg_set_bright bg_set_adjust bg_adjust_bright bg_adjust_color bg_adjust_ct bg_toggle dev_toggle', 'power': 'off', 'bright': '100', 'color_mode': '2', 'ct': '2801', 'rgb': '0', 'hue': '0', 'sat': '0', 'name': '', '_location_original': 'yeelight://192.168.107.53:55443', 'location': 'yeelight://192.168.107.53:55443', '_timestamp': datetime.datetime(2021, 12, 1, 6, 16, 41, 467365), '_host': '192.168.107.53', '_port': 49154, '_source': <SsdpSource.SEARCH: 'search'>}

```

bdraco commented:
Yeelight is a bit special because it runs SSDP discovery on an alternate port

https://github.com/home-assistant/core/blob/dev/homeassistant/components/yeelight/scanner.py#L157

epenet commented:
I can take a look tomorrow - feel free to add more information here in the meantime.
